### DISCUSSION TITLE
Events: Execute shell commands automatically when something happens

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#60] [Ideas]
This idea is about defining configurable events that can trigger shell commands to be executed. This would be one more step towards automation.

# Event ideas so far

- A bullet point item means: Idea (some of these are not yet researched for if they can be implemented or not)
- [ ] An empty box means: Will be implemented next.
- [x] A checked box means: Implemented, but not yet released.
- ✔️ means: Implemented and released.

## General
- ✔️ _After Obsidian starts_ (more specifically, this would actually be: _when the plugin is loaded_). This would allow a user to open a set of tool applications every time they open a certain Obsidian vault, e.g. [GIMP](https://www.gimp.org/) for manipulating note related images.
- ✔️ _Before Obsidian quits_.
- ✔️ *Every n seconds* (seconds could be later changed to a decimal number, allowing executing something e.g. every 0.5 seconds if needed)
- ✔️ *Command palette*: Currently this is the only action that can launch shell commands. It's included in this list to mark that a user can also turn it off, if they use other launchers for a shell command. Turn off = means that a shell command would not appear in command palette.

## Editor
- _When text is pasted_.  Would use a new [`editor-paste` workspace event](https://forum.obsidian.md/t/obsidian-release-v0-12-17-insider-build/25270). Could have an option to cancel the default pasting behavior and use the output channel _Current file: caret position_ to replace the pastable text with something else. This could e.g. allow doing custom formatting for pasted HTML or whatever.
- _When the caret is moved_. Note that this can happen very rapidly, so the shell command should not be anything heavy.
- _When a file is modified_. Note that this can happen very rapidly when a user is typing, so the shell command that is executed, should not be anything heavy. Would use a new [`editor-change` workspace event](https://forum.obsidian.md/t/obsidian-release-v0-12-17-insider-build/25270).

## File handling
- ✔️ _When a new file is created_.
- ✔️ _When a file is deleted_.
- ✔️ _When a file is moved_.
- ✔️ _When a file is renamed_.
- ✔️ _When a file is modified_.
- ✔️ _When a folder is created_.
- ✔️ _When a folder is renamed_.
- ✔️ _When a folder is deleted_.
- ✔️ _When a folder is moved_.
- _When external changes to files/folders are noticed_. (Edit 2022-05-09: The above events are triggered on external changes, too. Will have to see if there's any way to make a distinction between external and internal changes.)

## Special files / attachments
- _When a PDF is created_.
- _When an image is pasted from clipboard to Obsidian_. (This is a native feature in Obsidian, I'm not talking about the `{{clipboard}}` variable. Try to [search for an image on the web](https://duckduckgo.com/?q=image&t=hx&va=g&iax=images&ia=images), then right click an image, select *Copy image* or something (do not save the image as a file), then go to Obsidian, open some note in editor mode and simply paste, and you'll see a pasted image file appearing in your vault, with a link to it being inserted in the note).
- _When something is dropped to the editor_. I guess this would also handle the above image pasting thing? Would use a new [`editor-drop` workspace event](https://forum.obsidian.md/t/obsidian-release-v0-12-17-insider-build/25270).

## User interface
- ✔️ _After switching the active pane_. Example: Execute a shell command that reads the current file's last modification time and outputs the information to status bar (the status bar output feature is still under planning, see #16 ). The event helps to automatically give information for the file that is currently actively viewed in Obsidian.
- ~~_When the active note is closed_.~~ (Thanks Felipe for this idea!). Edit 2022-01-01: This will not be implemented, as _After switching the active pane_ will usually cover this situation, too. See this comment: https://github.com/Taitava/obsidian-shellcommands/discussions/60#discussioncomment-1800945
- ✔️ *File menu / Folder menu / Editor menu*: This would allow a user to define a new item in context menus. By clicking that context menu item, a shell command is executed, with possible event related variables like to what file was the context menu related to, or to what position in the editor.
- *When a link is clicked to open a file*: Would be able to get the link target file and alias text. Also wished to be able to prevent Obsidian from opening the file (optional). Thank you @DaveyUS for this idea! https://github.com/Taitava/obsidian-shellcommands/discussions/60#discussioncomment-1907879
- _When a window is opened/closed_ (`window-open` and `window-close`, [Obsidian 0.15.3](https://forum.obsidian.md/t/obsidian-release-v0-15-3-insider-build/39566)).

## Search
- _When something is searched_.

Note: I do not know yet, which one of these can be implemented in practice. Each event would need a similar event offered by Obsidian API, and I'm not familiar with what kind of events it provides. But ideas are always for good.

# Event related variables

Some events might need variables to provide data related to the event. As event related variables would not be available outside of events, ~~I'd like to introduce just one variable, `{{event}}`, so that a user clearly knows that this variable can only be used when a command is executed via an event. The `{{event}}` variable would have a parameter that would indicate the actual data that is needed:~~
- ~~`{{event:file_name}}` / `{{event:file_path}}`: The created/modified/deleted/moved file.~~
- ~~`{{event:search_keyword}}`~~
- ~~Something else?~~

Edit 2021-12-12: I think I'll create multiple variables for events, but they all will start with `{{event_`, e.g. `{{event_file_name}}`. Then it's easier to handle different sets of arguments, different help texts and different autocomplete items.

**Comment by FelipeRearden** [comment]
> Something else?
> When a user switches the active panel. 

1- When a user switches the active panel —> Save the caret position: line number and the count the amount of character:
- This could be a game changer when combined with [Obsidian Go To Line](https://github.com/phibr0/obsidian-go-to-line#1-go-to-line) plug-in.  
- Because Obsidian don’t save the character position after closing the note. 

2- When the file is closed —> Save the caret position: line number and the count the amount of character:
- This could be a game changer when combined with [Obsidian Go To Line](https://github.com/phibr0/obsidian-go-to-line#1-go-to-line) plug-in.  
- Because Obsidian don’t save the character position after closing the note.

**Reply by Taitava** [reply]
> 1- When a user switches the active panel —> Save the caret position: line number and the count the amount of character:

So that would be a use-case rather than something programmed into the plugin? The use-case could be supported by a variable that would give the caret position? That kind of variable idea could deserve it's own discussion, although I'm not yet very confident about how much use that variable would have.

> and the count the amount of character

Sorry, I didn't understand this part.

> 2- When the file is closed —> Save the caret position: line number and the count the amount of character:
>     * Because Obsidian don’t save the character position after closing the note.

Are you looking for a solution that when Obsidian starts, or a file gets reopened, Obsidian would place the caret at the same place where it was when the file was closed? In this case, I do not see a part in this workflow where actual shell commands would be used. To me it sounds more like a good feature request for Obsidian itself. 👍

**Reply by FelipeRearden** [reply]
Sorry for my bad English :) Let me try again 

> So that would be a use-case rather than something programmed into the plugin? The use-case could be supported by a variable that would give the caret position?

Yes, you are right! I was showing another example (Event + Example) of the event When the user switches the active panel  

My idea was to show more examples of how the [events +  Shell commands trigger] could be spectacular for complex workflows with other plugins involved. It generates a chain reactions that could be amazing!  

And a new event: When the active note is closed. 

- Event: When the user switches the active panel
- Trigger Shell Commands: run a Shell Command that stores (save) the line number and the character position inside the line [5:3 means line 5 and 3 characters from the beginning of the line]
- Benefits from other plugins: Using this feature with Obsidian Go-to Line
- User-case: store the caret position to use Obsidian Go-to line to create with complex workflows with QuickAdd

> Sorry, I didn't understand this part.

Same example with a new event: When the user closes the active note:

- New Event: When the user closes the active note;
- Trigger Shell Commands: run a Shell Command that stores (save) the line number and the character position inside the line [5:3 means line 5 and 3 characters from the beginning of the line];
- Benefits from other plugins: Using this feature with Obsidian Go-to Line;
- User-case: store the caret position to use Obsidian Go-to line to create with complex workflows with QuickAdd;

> Are you looking for a solution that when Obsidian starts, or a file gets reopened, Obsidian would place the caret at the same place where it was when the file was closed?

There is a plugin that have this feature :) [Obsidian Remember Cursor Position](https://github.com/derwish-pro/obsidian-remember-cursor-position/issues/7)
- and a FR made by me: I personally think that is not necessary to remember the caret position of every note. Only the notes that you choose.

—-

Every idea that you have @Taitava , triggers a new workflow in my head :)

**Reply by Taitava** [reply]
Thanks, now I understood your examples! 🙂 

> There is a plugin that have this feature :) [Obsidian Remember Cursor Position](https://github.com/derwish-pro/obsidian-remember-cursor-position/issues/7)

Good that a plugin exist, even when it's about a feature that should be (in my opinion) implemented in Obsidian itself in the future. But I know, Obsidian devs have hands full of work. 😎 

Here are a couple of FRs towards Obsidian:

- https://forum.obsidian.md/t/remember-position-when-switching-between-notes/1719
- https://forum.obsidian.md/t/restore-last-document-position-upon-starting-obsidian/962

>     * and a FR made by me: I personally think that is not necessary to remember the caret position of every note. Only the notes that you choose.

That's a good FR idea in the plugin!

**Reply by FelipeRearden** [reply]
Thanks @Taitava !!! 

I'm glad that we are on the same page!

> Here are a couple of FRs towards Obsidian:

Thanks foe these FRs !!!!!

**Reply by Taitava** [reply]
> 2- When the file is closed —> Save the caret position: line number and the count the amount of character:

I'd like to ask more about this particular idea. _When a file is closed_ can mean two things, I think: 1. Closing a pane that contains a file (happens rarely), 2. opening another file in a pane, which means that the previous file in the same pane is "closed", or at least not visible anymore.

I have now implemented _When a user switches the active panel_ * and I can see it has the following behavior:
 - It's triggered when:
   - user clicks a pane so that it becomes active (and it was not active before). Concerns whatever panes, e.g. a file or graph view or something else, but does not concern the left side menu (= clicking it does not remove the active color accent from the active pane).
   - user opens another file either to a new pane or current pane
   - user closes the currently active pane, in which case another pane becomes active, or a _No file is open_ pane open up and becomes active.
 - It's **not** triggered when:
   - a user closes a pane that was not active (click's precisely the X button of the pane, without first making the pane active)

**Summary and the actual question:** So, _When a user switches the active panel_ seems to be almost an "all around" event that covers even some cases of panel closing, althought not all, i.e. when an inactive pane is closed. Do you think there would still be a need for a _When a file is closed_ event? I tried to look it up quickly in Obsidian API, but didn't find anything like it at least yet. Or did you mean something else with _closing a file_?

*) Technically, _When a user switches the active panel_ uses [`active-leaf-change` Obsidian API event](https://github.com/obsidianmd/obsidian-api/blob/fac5e67f5d83829a4e0126905494c8cbca27765b/obsidian.d.ts#L3622). This is just for the record, and so that other programmers can possibly correct me if I'm wrong / provide adjustments.

**Reply by FelipeRearden** [reply]
@Taitava !!!!!

> Summary and the actual question: So, When a user switches the active panel seems to be almost an "all around" event that covers even some cases of panel closing, althought not all, i.e. when an inactive pane is closed. Do you think there would still be a need for a When a file is closed event? I tried to look it up quickly in Obsidian API, but didn't find anything like it at least yet.

I read 4 times and I understood perfectly. Following all you read I think we don't  need (see below):
> It's not triggered when:
a user closes a pane that was not active (click's precisely the X button of the pane, without first making the pane active)

Since we have all move related to switches to active panel, we have all covered (as you said)

> Or did you mean something else with closing a file?

No, my idea when close the panel is what you described here:
> user closes the currently active pane, in which case another pane becomes active, or a No file is open pane open up and becomes active.

---

This is absolutely amazing @Taitava ! I'm so happy to be part of this :)

Have a great day!

**Comment by Taitava** [comment]
I updated the first post in this discussion to include stuff related to [new workspace events](https://forum.obsidian.md/t/obsidian-release-v0-12-17-insider-build/25270) in Obsidian `0.12.17`.

**Comment by Taitava** [comment]
I might rename this *Events* feature to *Launchers* or something. It might be a bit more specific.

When these events/launchers will be implemented, I'll probably add *Command palette* as a launcher too, for the simple reason that then it can be turned off: a user could have shell commands that are not designed to be launched via the command palette. E.g. a shell command to show current file's size in statusbar could be launched when another file becomes active, but it does not need to unnecessarily fill command palette, which already has a huge list of commands.

I'll also add *Every n seconds* to the list of actions.

**Reply by FelipeRearden** [reply]
All great ideas @Taitava !!!! 

I like the idea of filter some SC commands from the Command Palette !!!!

**Reply by Taitava** [reply]
> I might rename this _Events_ feature to _Launchers_ or something. It might be a bit more specific.

I have tried to come up with a name that I would like for this feature, but I'm unsure.
- _Events_: Nice and short, but a bit ambiguous.
- _Launchers_: More specific, but will it indicate clearly that this is about automatisation? To me it sounds more like _"launcher icons"_ / _"launcher buttons"_.
- _Triggers_ / _Triggerers_: Bit complex sounding.

**Reply by FelipeRearden** [reply]
Actions
Routines
Flow 
Runners
Scripts
IFTTT
Connectors

**Reply by Taitava** [reply]
> Actions

I've been thinking about some kind of _"preactions"_, which would include at least the planned `{{input}}` variable. So preactions would be actions that will be done before shell command execution. That's why I keep _actions_ as a reserved term that should not be used for _events_.

> Runners

This brought to my mind: _executors_

> Scripts

To me _scripts_ sound more like files that contain multiple shell commands and that could be executed from the SC plugin (which is perfectly possible already).

> IFTTT

Kind of nice, as that's really what this feature will be about: _if this, then that_. But it's a brand name and a bit complex.

> Flow
> Connectors

Quite broad.

> Routines

Quite good one. I searched now for _command routines_ and found this page: https://support.google.com/googlenest/answer/7029585?gl=gb#Create%20a%20custom%20Routine (see: _Create a custom routine_ heading). There was also another term: _starters_

I know I'm overthinking this, I should just pick a term.

Edit: Thank you for your suggestions! 🙂

**Reply by FelipeRearden** [reply]
I loved Command Routines (CR) and Starters Routines (SR) or just Starters !!!!!!!! 

It's good to think about a name because we will talk about this features in many future FR and Discussions.  We need to find a name that make us comfortable :)

Have a great day @Taitava !!!!!!

Wait ..... Shell Routines might be a good one

**Reply by Taitava** [reply]
> It's good to think about a name because we will talk about this features in many future FR and Discussions. We need to find a name that make us comfortable :)

I think so too, and it's really important for new users: If a user doesn't recognize what a term means (i.e. for what a certain feature can be used for), the feature can be just a confusing looking button/tab in the userinterface, making the user hesitant to try it. A clear term can make the whole feature easier to understand and use.

> Wait ..... Shell Routines might be a good one

I think I'd like to avoid the term _shell_ here, as it's something that resides more on the operating system side, not so much on Obsidian's side. I haven't covered the term _shell_ very well in SC's documentation, would be nice to add something that covers what a _shell_ actually is. But here's something, although it might be complicated: https://en.wikipedia.org/wiki/Shell_(computing) . E.g. you have _Zsh (Z shell)_, that's one of the many _shells_ available.

Have a really nice day, too! 😎

**Reply by Taitava** [reply]
I think I'm going to keep the name _Events_, as JavaScript uses that name for the same thing.

**Comment by Taitava** [comment]
New idea added:
> - *Context menus*: This would allow a user to define a new item in context menus (editor context menu, file navigation context menu, what else are there?). By clicking that context menu item, a shell command is executed, with possible event related variables like to what file was the context menu related to, or to what position in the editor.

**Reply by FelipeRearden** [reply]
Great idea @Taitava !!!!!

**Comment by Taitava** [comment]
I added:
> - _When the caret is moved_.

**Comment by Taitava** [comment]
# Some progress...

Just to confirm shortly that I've been able to get a shell command executed when Obsidian opens and quits, and when active pane changes (= either another pane is clicked, or another file is opened).

**Reply by FelipeRearden** [reply]
That's amazing @Taitava !!!!!!! One step at time :)

Thanks for sharing this with me :)

Have a great day!!!!!!!!!!!!!!!!

**Reply by Taitava** [reply]
## Three menu events

Shell command aliases are used as menu item texts.

![kuva](https://user-images.githubusercontent.com/13002142/145953446-99c0e12b-a836-4f6f-8973-835c84483b1c.png)
Menu when right clicking a folder in the left side pane.

![kuva](https://user-images.githubusercontent.com/13002142/145953695-2ea9204f-d2bf-4a7c-8e24-ae375775e878.png)
Menu when right clicking a file in the left side pane.

![kuva](https://user-images.githubusercontent.com/13002142/145953809-e45228a4-035e-4a5e-9fc3-605bce6a92b4.png)
Menu when right clicking in editor. Note that this is the legacy editor, I'm still running Obsidian 0.12.19, so I don't have the new _Live preview_ editor that was introduced in Obsidian 0.13.0.

## Multiple events can be enabled for a single shell command
![kuva](https://user-images.githubusercontent.com/13002142/145954088-fcf04aa0-5146-4358-b22d-34480b7e2c61.png)
Plus shell commands can be hidden from Obsidian's command palette list.

**Reply by FelipeRearden** [reply]
Wooooowwwww @Taitava !!!!! 

I loved everything !!!! We have a new tab !!!!!!!!!!!!!!!!!!!

This is AMAZING @Taitava !!!! I can't believe it's happening :)

**Reply by Taitava** [reply]
Thank you a lot! 😎

I can't say yet when this will come out. At least not before Christmas. I think most of the coding for the first few events is done, but testing and documentation still needs to be done. After that I'll create another beta again.

**Reply by FelipeRearden** [reply]
> I can't say yet when this will come out. At least not before Christmas. I think most of the coding for the first few events is done, but testing and documentation still needs to be done. After that I'll create another beta again.

No problem at all @Taitava !!!! 

I'm just happy that we are in the road. Must be a difficult feature with many details that needs to be care :)

No rush :) 

Have a great day!

**Reply by Taitava** [reply]
I changed _After active leaf has changed_ to _After switching the active pane_. I think it's a bit more clearer. Obsidian API uses the term _leaf_, but Obsidian's UI uses the term _pane_.

The behavior is still the same.

**Comment by Taitava** [comment]
## A link click event?

@DaveyUS asked the following in https://github.com/Taitava/obsidian-shellcommands/issues/123#issuecomment-1004359568:
> What are the chances we could get an event that triggers when a link is clicked with a {{event_link_text}} variable?
> 
> A usage example would be when clicking [[audiofile.mp3]] the event would trigger allowing me pass the mp3 name to a python script.
> 
> Amazing plugin btw!

Thank you for your question and feedback! 🙂 At least _After switching the active pane_ event can be used for something close to what you are looking for (after the release of SC `0.10.0`). _After switching the active pane_ gets triggered every time a different file is opened in Obsidian, including `[[]]` links when they link to a file that Obsidian itself can open (so mp3 is supported). But that event can also be triggered by some other situations, e.g. user opening a file via the left side navigation, and when opening a graph view. The variables you would use with this event are the normal `{{file_name}}` etc. variables. No link text would be available when using this event. You could write some rules in your Python script to do stuff only if the file is mp3, if you want.

1. Do you need the link text especially? E.g. if you have a link like `[[audiofile.mp3|Some good music]]`, do you need to get the link alias `Some good music`? Or is it just enough to get the link's target file (`audiofile.mp3`)?
2. If only the file is needed, do you think _After switching the active pane_ could suffice your needs?
3. Would you like to prevent Obsidian from opening the link? So that the only thing that would happen, is to launch your Python script without opening the audio file in Obsidian?

I do not know Obsidian API completely, so I cannot say for all ideas if they are possible or not, but I like to have these conversations about things that could be nice to have, so that I can keep them on my mind and some day be able to tell if they can be implemented, even if I'm uncertain at the beginning. Hooking into link behavior is something that I'm not familiar with so far, but could be interesting.

**Reply by DaveyLovesCode** [reply]
Thanks for moving my question to the correct place!

To make it easier, I'll briefly explain my use case: I'm an avid voice note taker. It's the quickest way for me to get ideas out anytime, anywhere. I distill the voice notes down into written notes in Obsidian and leave links (with timestamps) to the original voice .mp3 so I can go back at anytime to hear more context if needed. I've built out a pretty nifty python setup that transcribes my voice into text (using the vosk speech library) allowing me to see a transcript of my words as I hear them. I can even click on my words as a way to seek/skip around the audio file. As a temporary solution, I've been linking the audio in obsidian and using the media-extended plugin as it allows for timestamps (voicefile.mp3#t=25 would play the file starting at 25 seconds), but that can only play the mp3 inside of obsidian.. it's completely disconnected from my python setup.

With that out of the way, here's answers to your questions:

1. I would need the entire link text in order to pass the timestamp off.
2. I suppose I'm not sure on this one. If it can pass the timestamp, then it could definitely work.
3. Ideally it would prevent obsidian from opening anything since python handles playing the audio.


In the end, all I need a shell command for is to feed the name and timestamp of the mp3 to my python system and it handles the rest. One could imagine other use cases where handing the link text to an external program on click could be useful. I really appreciate you taking the time to explore this idea!

**Reply by Taitava** [reply]
Thank you, it's a really interesting use case! 😎 

I'll add this to the list of ideas in the first post of this discussion, but my small research on Obsidian API did not reveal any way for me to access the clicked link. In other words, I don't currently know if this kind of link clicking event is possible to do at all. [I opened a discussion topic for it in Obsidian's forum](https://forum.obsidian.md/t/hook-into-link-click-event/30158). I can only continue planning this if the topic gets answered.

> 2\. I suppose I'm not sure on this one. If it can pass the timestamp, then it could definitely work.

As you do need the alias text (for the timestamp), _After switching the active pane_ is not a suitable event for this. AFAIK, there is no possibility to make that event able to retrive a clicked link's details. Also there's no possibility for this event to prevent opening the target file in Obsidian.

**Reply by DaveyLovesCode** [reply]
The level of dedication to your craft is inspiring. Thanks for taking the time to hear my idea out, and thanks for posting the request to the forums! Rock on 🤘

**Reply by Taitava** [reply]
# I'm moving the link click discussion again 🙂
@hakonhagland happened to post about a similar idea, so I decided that it's actually better to continue this discussion in the new discussion, where it's separate from all other events' discussions.

[If you have anything to comment about executing shell commands by clicking links/buttons in a note file, please comment in this other discussion](https://github.com/Taitava/obsidian-shellcommands/discussions/193).

**Comment by Taitava** [comment]
(Just for the record: beta version `0.10.0-beta.1` was made available a few days ago, so discussion from now on may relate to already available features.)

# A bug in _After switching the active pane_ event: It's sometimes executed twice

Sometimes, when switching the active pane, the event is triggered twice, and so all shell commands tied to the event, will also be executed twice. My current understanding is that this happens in Obsidian's side (Obsidian version 0.13.23), so it might be that there's not much I can do about it, but we'll see.

If this bug is problematic for your shell commands, you can read more details and watch a demonstration video from a bug report that I made on Obsidian's forum: https://forum.obsidian.md/t/api-active-leaf-change-event-is-sometimes-triggered-twice/31841

I will update this discussion when I get more information.
Edit 2022-04-03: The bug has been confirmed in the forum on 2022-02-04. I'll write an update again when it's fixed.

**Comment by Taitava** [comment]
# Part 2 implementation

The following events are now under development:
- When a new file is created.
- When a file is deleted.
- When a file is moved.
- When a file is renamed.
- When a folder is created.
- When a folder is renamed.
- When a folder is deleted.
- When a folder is moved.

Development process: #218

**Reply by Taitava** [reply]
These new events are now in beta test: https://github.com/Taitava/obsidian-shellcommands/discussions/228

**Reply by Taitava** [reply]
# Part 2 released
These events are now released in version `0.13.0`.

**Comment by Taitava** [comment]
# Window open/close related events

I added this to the list:

> - _When a window is opened/closed_ (`window-open` and `window-close`, [Obsidian 0.15.3](https://forum.obsidian.md/t/obsidian-release-v0-15-3-insider-build/39566)).

This is related to a new feature in Obsidian 0.15.x that allows locating panes in multiple windows.
